### PR TITLE
🌱 Changes to prepare for CAPI v1alpha4

### PR DESCRIPTION
**Pull request by fmuyassarov**
**What this PR does / why we need it**:
Changes required for provider,  for CAPI v1alpha4 release:

1. Update flag names to align with k8s components
   `--metrics-addr`  to  `--metrics-bind-address`
   `--leader-election`  to  `--leader-elect`
2. Upgrade kube-rbac-proxy to v0.8.0
3. Kustomize changes to have a single manager watching all namespaces and
answer to webhook calls.

To see the changes in CAPI from v1alpha3 -> v1alpha4, see the [doc](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md#align-manager-flag-names-with-upstream-kubernetes-components)

Fixes: https://github.com/metal3-io/cluster-api-provider-metal3/issues/160

### CODE REVIEW

**Review by kashifest — APPROVED** [review]

**Comment by kashifest** [comment]
/test-centos-integration
/test-v1a3-centos-integration
No logic to test these, but just to be sure ...

File: config/manager/manager_auth_proxy_patch.yaml [hunk 1]
```diff
@@ -21,5 +21,5 @@ spec:
           name: https
       - name: manager
         args:
-        - "--metrics-addr=127.0.0.1:8080"
-        - "--enable-leader-election"
+        - "--metrics-bind-address=127.0.0.1:8080"
```
Comment (furkatgofurov7) [inline config/manager/manager_auth_proxy_patch.yaml:24]:
```suggestion
        - "--metrics-bind-addr=127.0.0.1:8080"
```

File: config/webhook/manager_webhook_patch.yaml [hunk 2]
```diff
@@ -9,7 +9,7 @@ spec:
       containers:
       - name: manager
         args:
-          - "--metrics-addr=127.0.0.1:8080"
+          - "--metrics-bind-address=127.0.0.1:8080"
```
Comment (furkatgofurov7) [inline config/webhook/manager_webhook_patch.yaml:12]: same

File: main.go [hunk 3]
```diff
@@ -69,8 +69,8 @@ func init() {
 
 func main() {
 	klog.InitFlags(nil)
-	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
-	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
```
Comment (furkatgofurov7) [inline main.go:72]: same

**Review by furkatgofurov7 — CHANGES_REQUESTED** [review]
As per your patch description:

**Comment by fmuyassarov** [comment]
> As per your patch description:



> As per your patch description:

Updated the description.

**Comment by fmuyassarov** [comment]
@furkatgofurov7 reference document just in case
https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 reference document just in case
> https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/

Okay, this is a reference to kube-proxy btw, not kube-controller-manager.

**Comment by fmuyassarov** [comment]
> /test-v1a3-centos-integration

Seems we caught the problem 😀

**Comment by fmuyassarov** [comment]
> > @furkatgofurov7 reference document just in case
> > https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/
> 
> Okay, this is a reference to kube-proxy btw, not kube-controller-manager. I was checking the latter one initially.

you are right

**Comment by maelk** [comment]
You cannot use the v1a3 job on a PR that is towards CAPM3 master, as this is v1a4!

**Comment by kashifest** [comment]
> You cannot use the v1a3 job on a PR that is towards CAPM3 master, as this is v1a4!

Ah, true!

**Comment by kashifest** [comment]
The reason of failure is something else @fmuyassarov

**Comment by fmuyassarov** [comment]
> The reason of failure is something else @fmuyassarov

I think the reason is what @maelk mentioned. Because, I can see that there are two deployments of BMO, + CAPM3 is v1alpha4

```
[airshipci@ci-test-vm-20210209123434-gotb ~]$ kubectl get pods -A
NAMESPACE                           NAME                                                             READY   STATUS    RESTARTS   AGE
baremetal-operator-system           baremetal-operator-controller-manager-66b4fd8b65-56hkj           2/2     Running   0          29m
capi-kubeadm-bootstrap-system       capi-kubeadm-bootstrap-controller-manager-ddcc95784-6mlxf        2/2     Running   0          28m
capi-kubeadm-control-plane-system   capi-kubeadm-control-plane-controller-manager-d85dfd44c-4pg22    2/2     Running   0          28m
capi-system                         capi-controller-manager-5cd969bf55-mll24                         2/2     Running   0          28m
capi-webhook-system                 capi-controller-manager-68c9ff8646-pqqvz                         2/2     Running   0          28m
capi-webhook-system                 capi-kubeadm-bootstrap-controller-manager-7cf85756c5-kk249       2/2     Running   0          28m
capi-webhook-system                 capi-kubeadm-control-plane-controller-manager-7854964ff4-59m2w   2/2     Running   0          28m
capi-webhook-system                 capm3-controller-manager-6694579588-stvlp                        2/2     Running   0          28m
capi-webhook-system                 capm3-ipam-controller-manager-b75477967-6wm26                    2/2     Running   0          28m
capm3-system                        capm3-baremetal-operator-controller-manager-6b864d6b8d-d2thv     2/2     Running   0          28m
capm3-system                        capm3-controller-manager-5978456464-967px                        2/2     Running   0          28m
capm3-system                        capm3-ipam-controller-manager-5b4bd5c94b-xxlvz                   2/2     Running   0          28m
cert-manager                        cert-manager-cainjector-fc6c787db-2gj2h                          1/1     Running   0          28m
cert-manager                        cert-manager-d994d94d7-qp4zg                                     1/1     Running   0          28m
cert-manager                        cert-manager-webhook-845d9df8bf-968lw                            1/1     Running   0          28m
kube-system                         coredns-74ff55c5b-vdg6p                                          1/1     Running   0          29m
kube-system                         etcd-minikube                                                    1/1     Running   1          39m
kube-system                         kube-apiserver-minikube                                          1/1     Running   1          39m
kube-system                         kube-controller-manager-minikube                                 1/1     Running   1          39m
kube-system                         kube-proxy-ws8qm                                                 1/1     Running   0          29m
kube-system                         kube-scheduler-minikube                                          1/1     Running   1          39m
kube-system                         storage-provisioner                                              1/1     Running   1          39m
metal3                              metal3-ironic-5cb45dd9f6-82fmg                                   9/9     Running   0          28m

```



```
[airshipci@ci-test-vm-20210209123434-gotb ~]$ kubectl api-resources
NAME                              SHORTNAMES                APIVERSION                                 NAMESPACED   KIND
bindings                                                    v1                                         true         Binding
componentstatuses                 cs                        v1                                         false        ComponentStatus
configmaps                        cm                        v1                                         true         ConfigMap
endpoints                         ep                        v1                                         true         Endpoints
events                            ev                        v1                                         true         Event
limitranges                       limits                    v1                                         true         LimitRange
namespaces                        ns                        v1                                         false        Namespace
nodes                             no                        v1                                         false        Node
persistentvolumeclaims            pvc                       v1                                         true         PersistentVolumeClaim
persistentvolumes                 pv                        v1                                         false        PersistentVolume
pods                              po                        v1                                         true         Pod
podtemplates                                                v1                                         true         PodTemplate
replicationcontrollers            rc                        v1                                         true         ReplicationController
resourcequotas                    quota                     v1                                         true         ResourceQuota
secrets                                                     v1                                         true         Secret
serviceaccounts                   sa                        v1                                         true         ServiceAccount
services                          svc                       v1                                         true         Service
challenges                                                  acme.cert-manager.io/v1beta1               true         Challenge
orders                                                      acme.cert-manager.io/v1beta1               true         Order
clusterresourcesetbindings                                  addons.cluster.x-k8s.io/v1alpha3           true         ClusterResourceSetBinding
clusterresourcesets                                         addons.cluster.x-k8s.io/v1alpha3           true         ClusterResourceSet
mutatingwebhookconfigurations                               admissionregistration.k8s.io/v1            false        MutatingWebhookConfiguration
validatingwebhookconfigurations                             admissionregistration.k8s.io/v1            false        ValidatingWebhookConfiguration
customresourcedefinitions         crd,crds                  apiextensions.k8s.io/v1                    false        CustomResourceDefinition
apiservices                                                 apiregistration.k8s.io/v1                  false        APIService
controllerrevisions                                         apps/v1                                    true         ControllerRevision
daemonsets                        ds                        apps/v1                                    true         DaemonSet
deployments                       deploy                    apps/v1                                    true         Deployment
replicasets                       rs                        apps/v1                                    true         ReplicaSet
statefulsets                      sts                       apps/v1                                    true         StatefulSet
tokenreviews                                                authentication.k8s.io/v1                   false        TokenReview
localsubjectaccessreviews                                   authorization.k8s.io/v1                    true         LocalSubjectAccessReview
selfsubjectaccessreviews                                    authorization.k8s.io/v1                    false        SelfSubjectAccessReview
selfsubjectrulesreviews                                     authorization.k8s.io/v1                    false        SelfSubjectRulesReview
subjectaccessreviews                                        authorization.k8s.io/v1                    false        SubjectAccessReview
horizontalpodautoscalers          hpa                       autoscaling/v1                             true         HorizontalPodAutoscaler
cronjobs                          cj                        batch/v1beta1                              true         CronJob
jobs                                                        batch/v1                                   true         Job
kubeadmconfigs                                              bootstrap.cluster.x-k8s.io/v1alpha3        true         KubeadmConfig
kubeadmconfigtemplates                                      bootstrap.cluster.x-k8s.io/v1alpha3        true         KubeadmConfigTemplate
certificaterequests               cr,crs                    cert-manager.io/v1beta1                    true         CertificateRequest
certificates                      cert,certs                cert-manager.io/v1beta1                    true         Certificate
clusterissuers                                              cert-manager.io/v1beta1                    false        ClusterIssuer
issuers                                                     cert-manager.io/v1beta1                    true         Issuer
certificatesigningrequests        csr                       certificates.k8s.io/v1                     false        CertificateSigningRequest
clusters                          cl                        cluster.x-k8s.io/v1alpha3                  true         Cluster
machinedeployments                md                        cluster.x-k8s.io/v1alpha3                  true         MachineDeployment
machinehealthchecks               mhc,mhcs                  cluster.x-k8s.io/v1alpha3                  true         MachineHealthCheck
machines                          ma                        cluster.x-k8s.io/v1alpha3                  true         Machine
machinesets                       ms                        cluster.x-k8s.io/v1alpha3                  true         MachineSet
providers                                                   clusterctl.cluster.x-k8s.io/v1alpha3       true         Provider
kubeadmcontrolplanes              kcp                       controlplane.cluster.x-k8s.io/v1alpha3     true         KubeadmControlPlane
leases                                                      coordination.k8s.io/v1                     true         Lease
endpointslices                                              discovery.k8s.io/v1beta1                   true         EndpointSlice
events                            ev                        events.k8s.io/v1                           true         Event
machinepools                      mp                        exp.cluster.x-k8s.io/v1alpha3              true         MachinePool
ingresses                         ing                       extensions/v1beta1                         true         Ingress
flowschemas                                                 flowcontrol.apiserver.k8s.io/v1beta1       false        FlowSchema
prioritylevelconfigurations                                 flowcontrol.apiserver.k8s.io/v1beta1       false        PriorityLevelConfiguration
metal3clusters                    m3c,m3cluster             infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3Cluster
metal3dataclaims                  m3dc,m3dataclaim          infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3DataClaim
metal3datas                       m3d,m3data                infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3Data
metal3datatemplates               m3dt,m3datatemplate       infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3DataTemplate
metal3machines                    m3m,m3machine             infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3Machine
metal3machinetemplates                                      infrastructure.cluster.x-k8s.io/v1alpha4   true         Metal3MachineTemplate
ipaddresses                       ipa,ipaddress,ipaddress   ipam.metal3.io/v1alpha1                    true         IPAddress
ipclaims                          m3ipc,m3ipclaim,ipclaim   ipam.metal3.io/v1alpha1                    true         IPClaim
ippools                           m3ipp,m3ippool            ipam.metal3.io/v1alpha1                    true         IPPool
baremetalhosts                    bmh,bmhost                metal3.io/v1alpha1                         true         BareMetalHost
ingressclasses                                              networking.k8s.io/v1                       false        IngressClass
ingresses                         ing                       networking.k8s.io/v1                       true         Ingress
networkpolicies                   netpol                    networking.k8s.io/v1                       true         NetworkPolicy
runtimeclasses                                              node.k8s.io/v1                             false        RuntimeClass
poddisruptionbudgets              pdb                       policy/v1beta1                             true         PodDisruptionBudget
podsecuritypolicies               psp                       policy/v1beta1                             false        PodSecurityPolicy
clusterrolebindings                                         rbac.authorization.k8s.io/v1               false        ClusterRoleBinding
clusterroles                                                rbac.authorization.k8s.io/v1               false        ClusterRole
rolebindings                                                rbac.authorization.k8s.io/v1               true         RoleBinding
roles                                                       rbac.authorization.k8s.io/v1               true         Role
priorityclasses                   pc                        scheduling.k8s.io/v1                       false        PriorityClass
csidrivers                                                  storage.k8s.io/v1                          false        CSIDriver
csinodes                                                    storage.k8s.io/v1                          false        CSINode
storageclasses                    sc                        storage.k8s.io/v1                          false        StorageClass
volumeattachments                                           storage.k8s.io/v1                          false        VolumeAttachment
```

**Comment by kashifest** [comment]
> I think the reason is what @maelk mentioned. Because, I can see that there are two deployments of BMO, + CAPM3 is v1alpha4
Yes I got it now

**Pull request closed without merging**